### ISSUE TITLE
sql: enum types in CHECK constraints are not human readable in SHOW CONSTRAINTS, pg_catalog.pg_constraint, and pg_get_constraintdef

### ISSUE THREAD

**Issue by mgartner** [cockroachdb/cockroach#53918]
Enums in CHECK constraints are not human readable the `condef` column of `pg_catalog.pg_constraint`. As a result, they are also not human reaable in the output of `SHOW CONSTRAINTS` and `pg_get_constraintdef`.

Note that `pgcatalog.pg_constraint.conbin` does not have to be human readable—it is not human readable in Postgres. Postgres has not `pgcatalog.pg_constraint.condef` column.

### CockroachDB

```
root@127.0.0.1:53152/defaultdb> create type e as enum ('foo', 'bar', 'baz');
CREATE TYPE


root@127.0.0.1:53152/defaultdb> create table t (a e, check (a IN ('foo', 'bar')));
CREATE TABLE


root@127.0.0.1:53152/defaultdb> show constraints from t;
  table_name | constraint_name | constraint_type |                        details                        | validated
-------------+-----------------+-----------------+-------------------------------------------------------+------------
  t          | check_a         | CHECK           | CHECK ((a IN (b'\x40':::@100052, b'\x80':::@100052))) |   true


root@127.0.0.1:53152/defaultdb> select oid, conname, conbin, consrc, condef from pg_catalog.pg_constraint;
     oid    | conname |                    conbin                     |                    consrc                     |                        condef
------------+---------+-----------------------------------------------+-----------------------------------------------+--------------------------------------------------------
  165418667 | check_a | (a IN (b'\x40':::@100052, b'\x80':::@100052)) | (a IN (b'\x40':::@100052, b'\x80':::@100052)) | CHECK ((a IN (b'\x40':::@100052, b'\x80':::@100052)))


root@127.0.0.1:53152/defaultdb> select pg_get_constraintdef(165418667);
                  pg_get_constraintdef
---------------------------------------------------------
  CHECK ((a IN (b'\x40':::@100052, b'\x80':::@100052)))
```

### In Postgres

```
marcus=# create type e as enum ('foo', 'bar', 'baz');
CREATE TYPE


marcus=# create table t (a e, check (a IN ('foo', 'bar')));
CREATE TABLE


marcus=# select oid, conname, conbin from pg_catalog.pg_constraint where conname = 't_a_check';
  oid  |  conname  |                                                                                                                                                                                                                                                                                                                         conbin
-------+-----------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 74862 | t_a_check | {SCALARARRAYOPEXPR :opno 3516 :opfuncid 3508 :useOr true :inputcollid 0 :args ({VAR :varno 1 :varattno 1 :vartype 74853 :vartypmod -1 :varcollid 0 :varlevelsup 0 :varnoold 1 :varoattno 1 :location 28} {ARRAY :array_typeid 74852 :array_collid 0 :element_typeid 74853 :elements ({CONST :consttype 74853 :consttypmod -1 :constcollid 0 :constlen 4 :constbyval true :constisnull false :location 34 :constvalue 4 [ 102 36 1 0 0 0 0 0 ]} {CONST :consttype 74853 :consttypmod -1 :constcollid 0 :constlen 4 :constbyval true :constisnull false :location 41 :constvalue 4 [ 104 36 1 0 0 0 0 0 ]}) :multidims false :location -1}) :location 30}


marcus=# select pg_get_constraintdef(74862);
             pg_get_constraintdef
-----------------------------------------------
 CHECK ((a = ANY (ARRAY['foo'::e, 'bar'::e])))
```

**Comment by ajwerner** [comment]
I can take a look at this when I get home in a bit. I think this will happen if the types are not properly hydrated.

**Comment by ajwerner** [comment]
This diff would do it, I don't know if it's advisable:

```go
diff --git a/pkg/sql/pg_catalog.go b/pkg/sql/pg_catalog.go
index 3a90f9f784..c7ef63768e 100644
--- a/pkg/sql/pg_catalog.go
+++ b/pkg/sql/pg_catalog.go
@@ -978,7 +978,7 @@ func populateTableConstraints(
                        if conkey, err = colIDArrayToDatum(con.CheckConstraint.ColumnIDs); err != nil {
                                return err
                        }
-                       displayExpr, err := schemaexpr.FormatExprForDisplayWithoutTypeAnnotations(ctx, table, con.Details, &p.semaCtx)
+                       displayExpr, err := schemaexpr.FormatExprForDisplay(ctx, table, con.Details, &p.semaCtx)
                        if err != nil {
                                return err
                        }
```

**Comment by mgartner** [comment]
@ajwerner thanks for taking a look. 

The problem with your diff is that for `pg_catalog` values we cannot have `:::` annotations for capatibility with ORMs. See: https://github.com/cockroachdb/cockroach/pull/41323/files#diff-927b5724fbec50a5e9dc254a7dc27fe9R120

I'm assuming having `:::` annotations in the output of `pg_get_constraintdef` would also be problematic.

There's a deeper issue around formatting here, given the number of distinct code paths we have for formatting. Without unified formatting logic, its easy for someone to miss places that need to be updated. Like I have done myself: https://github.com/cockroachdb/cockroach/issues/53908 https://github.com/cockroachdb/cockroach/issues/53843

I'm sure those aren't the only two examples either.

**Comment by mgartner** [comment]
And I'm now regretting putting some of the formatting logic in `schemaexpr`. I don't think it belongs there and is only adding to the confusion. I'd like to tidy things up, but for the imminent release we'll have to settle for some quick-fixes.

**Comment by mgartner** [comment]
Doh. This was fixed by #53865.
